[Tan Wei Howe] iP

whtan99 opened the pull request:
<h1>Duke</h1>

> “Your mind is for having ideas, not holding them.” – David Allen [Source](https://dansilvestre.com/productivity-quotes)

Let Duke handle remembering the things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ *SUPER* FAST to use.

All you need to do is,
1. download it from [here](https://github.com/goalfix/ip/releases/tag/A-Jar)
2. double-click on it.
3. add your tasks.
4. let it manage your tasks for you :smiley:

The best part - ITS **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java Programmer, you can use it to practice Java too! Here's the `main` method, found within the `Duke` class:
```java
public static void main(String[] args) {
    new Duke("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Commit:
Level 1 greet echo bye working

Commit:
add, list task working

Commit:
mark unmark task done, to refactor code

Commit:
refactor code and rename methods to follow java conventions

Commit:
added tracking for three types of tasks

Commit:
added script for automated testing of level 4

Commit:
deal with errors such as incorrect inputs by user

Commit:
Add functionality to delete tasks

Commit:
add ability to save files locally

Commit:
add datetime functionality

Commit:
Merge branch 'branch-Level-8'

* branch-Level-8:
  add datetime functionality

# Conflicts:
#	src/main/java/Duke.java

Commit:
refactor code into OOP style

Commit:
handle exceptions after using OOP style

Commit:
put all classes into Duke package

Commit:
rename packages

Commit:
add junit test

Commit:
reorganise main directory

Commit:
update gitignore to prevent tracking of MANIFEST file

Commit:
add javadocs

Commit:
fix sytle violations

Commit:
add ability to find by keyword

Commit:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  add javadocs

Commit:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  fix sytle violations

Commit:
Merge branch 'branch-Level-9'

* branch-Level-9:
  add ability to find by keyword

Commit:
Merge branch 'add-gradle-support' of https://github.com/goalfix/ip

Commit:
add gradle support to build and run duke

Commit:
ensure duke following coding conventions

Commit:
add GUI, WIP

waffledood commented on src/main/java/duke/AddTaskCommand.java:
```diff
@@ -0,0 +1,74 @@
+                tasks.addTask(new Deadline(description, toDoBy));
+            } else {
+                tasks.addTask(new Event(description, toDoBy));
+            }
```
That's good you used `AddCommand` to house ToDo, Deadline & Event tasks, that's what I did as well!
Perhaps you can consider using a `switch` statement for this part so it looks more clean? 

```suggestion
            switch (type) {
            case "todo":
                tasks.addTask(new ToDo(description));
                break; 
            case "deadline": 
                tasks.addTask(new Deadline(description, toDoBy));
                break;
            case "event":
                tasks.addTask(new Event(description, toDoBy));
                break;
            default:
                ui.showError("Unexpected command");
                break;
            }
```

waffledood commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,61 @@
+            }
+        case ("deadline"):
+            try {
+                String[] descriptionsSplitByTime = descriptions[1].split(" /by ", 2);
```
Consider using `"/by"` instead. If the user inputs `deadline return book/by2021-01-01 01:01`, the `String[]` array will instead be `String[1] { "return book/by2021-01-01 01:01" }`

```suggestion
                String[] descriptionsSplitByTime = descriptions[1].split("/by", 2);
```

waffledood commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,61 @@
+            }
+        case ("event"):
+            try {
+                String[] descriptionsSplitAtTime = descriptions[1].split(" /at ", 2);
```
Similarly with `"/at"` as `"/by"`

```suggestion
                String[] descriptionsSplitAtTime = descriptions[1].split("/at", 2);
```

waffledood commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,55 @@
+     *
+     * @return Task in file format.
+     */
+    public String toFileFormat() {
+        return null;
+    }
```
Instead of having `toFileFormat()` return `null`, the `Task` class can be implemented as an `abstract` class & the `toFileFormat()` as an `abstract` method. 

```suggestion
    public abstract String toFileFormat();
```

waffledood commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,55 @@
+package main.java.duke;
+
+
+public class Task {
```
Instead of having `toFileFormat()` return `null`, the `Task` class can be implemented as an `abstract` class & the `toFileFormat()` as an `abstract` method. 

This wouldn't affect your current implementation with `ToDo`, `Deadline` & `Event` already extending `Task`

```suggestion
public abstract class Task {
```

waffledood commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,80 @@
+
+public class TaskList {
+    private ArrayList<Task> taskArr;
+    private int currTask;
```
I suppose `currTask` is used to track the number of tasks & thus size of the `TaskList`? But in the method `size()`, `taskArr.size()` was called, so there wouldn't be a need for `currTask` then?
```suggestion
```

waffledood commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,80 @@
+    /** Constructs an instance of Tasklist. */
+    public TaskList() {
+        this.taskArr = new ArrayList<Task>();
+        currTask = 0;
```
```suggestion

```

waffledood commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,80 @@
+     */
+    public TaskList(ArrayList<Task> t) {
+        this.taskArr = t;
+        this.currTask = t.size();
```
```suggestion

```

waffledood commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,80 @@
+     */
+    public void addTask(Task t) {
+        taskArr.add(t);
+        currTask++;
```
```suggestion

```

waffledood commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,80 @@
+     */
+    public void delete(int taskNum) {
+        taskArr.remove(taskNum);
+        currTask--;
```
```suggestion

```

waffledood commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,161 @@
+     *
+     * @param type Type of error.
+     */
+    public void showError(String type) {
+        switch (type) {
+        case ("LoadingError"):
+            System.out.println("Something went wrong with the loading of the file");
+            System.out.println("Try deleting the data/tasks.txt file, but this will "
+                    + "wipe all previous tasks stored (if any).");
+            break;
+        case ("UnknownCommand"):
+            System.out.println("I'm sorry, but I can't execute this command!");
+            break;
+        case ("TodoFormatError"):
+            System.out.println("Follow this format:");
+            System.out.println("todo YOUR_TASK");
+            break;
+        case ("DeadlineFormatError"):
+            System.out.println("Follow this format:");
+            System.out.println("deadline YOUR_TASK /by yyyy-mm-dd TIME");
+            System.out.println("eg: deadline return book /by 2019-10-15 18:00");
+            break;
+        case ("EventFormatError"):
+            System.out.println("Follow this format:");
+            System.out.println("event YOUR_TASK /at yyyy-mm-dd TIME");
+            System.out.println("eg: event project meeting /at 2019-10-15 18:00");
+            break;
+        case ("IOException"):
+            System.out.println("Something went wrong with the writing to the file");
+            break;
+        case ("DateTimeParseException"):
+            System.out.println("Please enter the correct format for Datetime! yyyy-mm-dd HH:mm");
+            break;
+        default:
+            System.out.println("I don't understand this error!");
+        }
```
Wow, very detailed, great job! 👍 🚀

waffledood commented on src/test/java/duke/TaskTest.java:
```diff
@@ -0,0 +1,18 @@
+package duke;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Separation of import statements as per [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements) 

```suggestion
package duke;

import static org.junit.jupiter.api.Assertions.assertEquals;
```

waffledood commented on src/test/java/duke/EventTest.java:
```diff
@@ -0,0 +1,16 @@
+package duke;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Separation of package statements as per [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)

```suggestion
package duke;

import static org.junit.jupiter.api.Assertions.assertEquals;
```

waffledood commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,87 @@
+package main.java.duke;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Scanner;
```
Separation of package statements as per [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)

```suggestion
package main.java.duke;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;

import java.nio.file.Paths;

import java.util.ArrayList;
import java.util.Scanner;
```

waffledood commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,161 @@
+package main.java.duke;
+import java.util.ArrayList;
+import java.util.Scanner;
```
Separation of package statements as per [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)

```suggestion
package main.java.duke;

import java.util.ArrayList;
import java.util.Scanner;
```

waffledood commented on src/main/java/duke/UnmarkCommand.java:
```diff
@@ -0,0 +1,44 @@
+package main.java.duke;
+import java.io.IOException;
```
Separation of package statements as per [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)

```suggestion
package main.java.duke;

import java.io.IOException;
```

waffledood commented on src/main/java/duke/MarkCommand.java:
```diff
@@ -0,0 +1,39 @@
+package main.java.duke;
+import java.io.IOException;
```
Separation of package statements as per [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)

```suggestion
package main.java.duke;

import java.io.IOException;
```

waffledood commented on src/main/java/duke/ListCommand.java:
```diff
@@ -0,0 +1,34 @@
+package main.java.duke;
+import java.util.ArrayList;
```
Separation of package statements as per [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)

```suggestion
package main.java.duke;

import java.util.ArrayList;
```

waffledood commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,41 @@
+package main.java.duke;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
```
Separation of package statements as per [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)
```suggestion
package main.java.duke;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
```

waffledood commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+package main.java.duke;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
```
Separation of package statements as per [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)
```suggestion
package main.java.duke;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
```

waffledood commented on src/main/java/duke/AddTaskCommand.java:
```diff
@@ -0,0 +1,74 @@
+package main.java.duke;
+
+import java.io.IOException;
+import java.time.format.DateTimeParseException;
```
Separation of package statements as per [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)
```suggestion
package main.java.duke;

import java.io.IOException;

import java.time.format.DateTimeParseException;
```

waffledood reviewed (commented):
LGTM overall! 
Just a few cosmetic changes suggested (in line with Importing Packages [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements))

Commit:
add gui support

Commit:
Merge branch 'add-gui'

* add-gui:
  add gui support
  add GUI, WIP

Commit:
fix error handling for gui

Commit:
Merge branch 'branch-Level-10'

* branch-Level-10:
  fix error handling for gui

Commit:
Add assertions

Currently, no assertions to make sure critical assumptions hold.

Assertions are needed to ensure that there are no potential bugs in the code.

Let's,
* add assertion in AddTaskCommand, FindCommand, Task to ensure description is not empty
* add assertion in Storage to ensure that a filepath is provided

Unused code is being committed out as well.

Commit:
Improve Quality of Code

Currently, some code is unused.

Dead code should be removed, to remove excess confusion to user reading the program.

Let's,
* remove the unused constructor in UnmarkCommand
* remove the unused showKeywords method in Ui

whtan99 committed:
Merge pull request #1 from goalfix/branch-A-Assertions

Add assertions

Commit:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assertions

whtan99 committed:
Merge pull request #2 from goalfix/branch-A-CodeQuality

Branch a code quality

Commit:
add functionality to prevent duplicate task from being added

Commit:
Merge branch 'branch-C-DetectDuplicates'

* branch-C-DetectDuplicates:
  add functionality to prevent duplicate task from being added

Commit:
add representative screenshot

whtan99 committed:
Update ReadMe.md

whtan99 committed:
add user guide

Commit:
fix coding standards

Commit:
Merge branch 'master' of https://github.com/goalfix/ip

Commit:
fix checkstyle errors
